File: algos/solutions/a1twopointers/TS/threesum/algo.test.ts
import { threeSum } from "./algo";

// Helper function to sort triplets for consistent comparison.
function sortTriplets(triplets: number[][]): number[][] {
    return triplets
        .map(triplet => triplet.slice().sort((a, b) => a - b))
        .sort((a, b) => {
            for (let i = 0; i < a.length; i++) {
                if (a[i] !== b[i]) return a[i] - b[i];
            }
            return 0;
        });
}

describe("threeSum", () => {
    test("should return all unique triplets that sum to zero", () => {
        const nums = [-1, 0, 1, 2, -1, -4];
        const expected = [[-1, -1, 2], [-1, 0, 1]];
        const result = threeSum(nums);
        expect(sortTriplets(result)).toEqual(sortTriplets(expected));
    });

    test("should return an empty array when no triplet sums to zero", () => {
        const nums = [1, 2, 3, 4, 5];
        expect(threeSum(nums)).toEqual([]);
    });

    test("should handle cases with multiple zeros", () => {
        const nums = [0, 0, 0, 0];
        const expected = [[0, 0, 0]];
        expect(threeSum(nums)).toEqual(expected);
    });
});

File: algos/solutions/a1twopointers/Go/threesum/algo_test.go
package threesum

import (
	"reflect"
	"sort"
	"testing"
)

// sortTriplets sorts each triplet and then sorts the list of triplets for consistent comparison.
func sortTriplets(triplets [][]int) [][]int {
	for _, triplet := range triplets {
		sort.Ints(triplet)
	}
	sort.Slice(triplets, func(i, j int) bool {
		for k := 0; k < len(triplets[i]); k++ {
			if triplets[i][k] != triplets[j][k] {
				return triplets[i][k] < triplets[j][k]
			}
		}
		return false
	})
	return triplets
}

func TestThreeSum(t *testing.T) {
	tests := []struct {
		input    []int
		expected [][]int
	}{
		{
			input:    []int{-1, 0, 1, 2, -1, -4},
			expected: [][]int{{-1, -1, 2}, {-1, 0, 1}},
		},
		{
			input:    []int{1, 2, 3, 4, 5},
			expected: [][]int{},
		},
		{
			input:    []int{0, 0, 0, 0},
			expected: [][]int{{0, 0, 0}},
		},
	}

	for _, tt := range tests {
		got := ThreeSum(tt.input)
		if !reflect.DeepEqual(sortTriplets(got), sortTriplets(tt.expected)) {
			t.Errorf("ThreeSum(%v) = %v; want %v", tt.input, got, tt.expected)
		}
	}
}
